### PR TITLE
RGW | Bucket Notification: migrating old entries to support persistency control

### PR DESCRIPTION
**Pull request by AliMasarweh**
The way I tested the migration is by making a commit that simulates old gen entries (old gen entries don't have a creation timestamp) then kill rgw and build new bin for it then again re-ran the cluster
```
-      event_entry.creation_time = ceph::coarse_real_clock::now();
+      event_entry.creation_time = ceph::coarse_real_clock::zero();
```

now we build and run the cluster with a topic that configures time to live to be 10 seconds:
```
$ aws --endpoint-url http://localhost:8000 sns create-topic --name=fishtopic \
	--attributes='{"push-endpoint": "http://localhost:10900", "persistent": "True", "time_to_live": "10" }'
```
we verify that after 10 seconds entries are still there:
```
$ bin/radosgw-admin topic stats --topic fishtopic
{
    "Topic Stats": {
        "Reservations": 0,
        "Size": 7074,
        "Entries": 12
    }
}
```
then we proceed to migration:
```
$ pgrep -a radosgw
678785 /home/alimasa/ceph/build/bin/radosgw -c /home/alimasa/ceph/build/ceph.conf --log-file=/home/alimasa/ceph/build/out/radosgw.8000.log --admin-socket=/home/alimasa/ceph/build/out/radosgw.8000.asok --pid-file=/home/alimasa/ceph/build/out/radosgw.8000.pid --rgw_luarocks_location=/home/alimasa/ceph/build/out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends=beast port=8000
681199 ./bin/radosgw-admin bucket stats --bucket=test-100m-1000000000000 --sync-stats
681200 ./bin/radosgw-admin bucket stats --bucket=test-100m-1000000000000 --sync-stats
$ git reset --hard HEAD~1
$ git cherry-pick COMMIT_HASH # this PR hash
$ ninja
$ /home/alimasa/ceph/build/bin/radosgw -c /home/alimasa/ceph/build/ceph.conf --log-file=/home/alimasa/ceph/build/out/radosgw.8000.log --admin-socket=/home/alimasa/ceph/build/out/radosgw.8000.asok --pid-file=/home/alimasa/ceph/build/out/radosgw.8000.pid --rgw_luarocks_location=/home/alimasa/ceph/build/out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends="beast port=8000"
```
then we wait until RGW is back online
```
$ pgrep -a radosgw
683428 /home/alimasa/ceph/build/bin/radosgw -c /home/alimasa/ceph/build/ceph.conf --log-file=/home/alimasa/ceph/build/out/radosgw.8000.log --admin-socket=/home/alimasa/ceph/build/out/radosgw.8000.asok --pid-file=/home/alimasa/ceph/build/out/radosgw.8000.pid --rgw_luarocks_location=/home/alimasa/ceph/build/out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends=beast port=8000
684210 ./bin/radosgw-admin sync error list
684211 ./bin/radosgw-admin sync error list
$ pgrep -a radosgw
683428 /home/alimasa/ceph/build/bin/radosgw -c /home/alimasa/ceph/build/ceph.conf --log-file=/home/alimasa/ceph/build/out/radosgw.8000.log --admin-socket=/home/alimasa/ceph/build/out/radosgw.8000.asok --pid-file=/home/alimasa/ceph/build/out/radosgw.8000.pid --rgw_luarocks_location=/home/alimasa/ceph/build/out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends=beast port=8000
```
then we wait another 10 seconds and all the entries will expire:
```
$ bin/radosgw-admin topic stats --topic fishtopic
{
    "Topic Stats": {
        "Reservations": 0,
        "Size": 0,
        "Entries": 0
    }
}

```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/cls/queue/cls_queue_types.h [hunk 1]
```diff
@@ -20,18 +20,23 @@ struct cls_queue_entry
 {
   ceph::buffer::list data;
   std::string marker;
+  bool needs_migration = false;
 
   void encode(ceph::buffer::list& bl) const {
     ENCODE_START(1, 1, bl);
```
Comment (yuvalif) [inline src/cls/queue/cls_queue_types.h:26]: you should change the version to "2", with "1" as compatible

File: src/rgw/driver/rados/rgw_notify.cc [hunk 2]
```diff
@@ -28,9 +28,10 @@ struct event_entry_t {
   uint32_t time_to_live = DEFAULT_GLOBAL_VALUE;
   uint32_t max_retries = DEFAULT_GLOBAL_VALUE;
   uint32_t retry_sleep_duration = DEFAULT_GLOBAL_VALUE;
+  bool needs_migration = false;
   
   void encode(bufferlist& bl) const {
-    ENCODE_START(3, 1, bl);
+    ENCODE_START(4, 1, bl);
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:34]: if we changed from "2" to "3" post "reef", you don't need to change to "4" - you can keep at "3".
there are no compatibility requirements within the same version (both changes are in "squid").

File: src/rgw/driver/rados/rgw_notify.cc [hunk 3]
```diff
@@ -28,9 +28,10 @@ struct event_entry_t {
   uint32_t time_to_live = DEFAULT_GLOBAL_VALUE;
   uint32_t max_retries = DEFAULT_GLOBAL_VALUE;
   uint32_t retry_sleep_duration = DEFAULT_GLOBAL_VALUE;
+  bool needs_migration = false;
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:31]: why not init to "true"?

File: src/rgw/driver/rados/rgw_notify.cc (hunk 2 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc:34]: Wasn't aware that it is tied to the ceph versions
Thanks for informing me :)

File: src/rgw/driver/rados/rgw_notify.cc (hunk 3 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc:31]: I was doing it in the else bracket if it didn't decode, but this way is optimal

File: src/test/rgw/bucket_notification/test_bn.py [hunk 4]
```diff
@@ -3262,8 +3262,8 @@ def test_ps_s3_persistent_topic_configs_ttl():
 @attr('basic_test')
 def test_ps_s3_persistent_topic_configs_max_retries():
     """ test persistent topic configurations with max_retries and retry_sleep_duration """
-    config_dict = {"time_to_live": "None", "max_retries": 20, "retry_sleep_duration": 1}
-    buffer = 20
+    config_dict = {"time_to_live": "None", "max_retries": 10, "retry_sleep_duration": 1}
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:3265]: how are these changes related?

File: src/rgw/driver/rados/rgw_notify.cc [hunk 5]
```diff
@@ -206,6 +209,11 @@ class Manager : public DoutPrefixProvider {
       return EntryProcessingResult::Failure;
     }
 
+    if (event_entry.needs_migration) {
+      tenant_name = event_entry.event.userIdentity;
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:213]: are you sure that the user identity is the tenant?

File: src/cls/queue/cls_queue_types.h [hunk 6]
```diff
@@ -20,18 +20,23 @@ struct cls_queue_entry
 {
   ceph::buffer::list data;
   std::string marker;
+  bool needs_migration = false;
```
Comment (yuvalif) [inline src/cls/queue/cls_queue_types.h:23]: why do we need to change the cls_queue entry type?
this is an abstract queue, unrelated to notifications

File: src/rgw/driver/rados/rgw_notify.cc [hunk 7]
```diff
@@ -28,6 +28,7 @@ struct event_entry_t {
   uint32_t time_to_live = DEFAULT_GLOBAL_VALUE;
   uint32_t max_retries = DEFAULT_GLOBAL_VALUE;
   uint32_t retry_sleep_duration = DEFAULT_GLOBAL_VALUE;
+  bool needs_migration = true;
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:31]: can we use `creation_time == ceph::coarse_real_clock::zero()` as an indication that mogration is needed, instead of adding a new data member?

File: src/rgw/driver/rados/rgw_notify.cc [hunk 8]
```diff
@@ -381,23 +390,31 @@ class Manager : public DoutPrefixProvider {
...
             auto& persistency_tracker = notifs_persistency_tracker[entry.marker];
-            auto result = process_entry(this->get_cct()->_conf, persistency_tracker, entry, yield);
-            if (result == EntryProcessingResult::Successful || result == EntryProcessingResult::Expired) {
-              ldpp_dout(this, 20) << "INFO: processing of entry: " << 
-                entry.marker << " (" << entry_idx << "/" << total_entries << ") from: " << queue_name
-                << (result == EntryProcessingResult::Successful? " ok": " expired") << dendl;
+            auto result = process_entry(this->get_cct()->_conf, persistency_tracker, entry, tenant_name, yield);
+            if (result == EntryProcessingResult::Successful || result == EntryProcessingResult::Expired
+                || result == EntryProcessingResult::Migrating) {
+              std::string return_status = "ok";
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:399]: if `return_status` is only used in the debug log, please avoid setting it here.
in line 408 you can just log the enum, or convert the enum to string there, so it is not calculated if debug level is < 20

File: src/rgw/driver/rados/rgw_notify.cc [hunk 9]
```diff
@@ -440,9 +462,67 @@ class Manager : public DoutPrefixProvider {
...
             << queue_name << ". error: " << ret << dendl;
+          return;
         } else {
-          ldpp_dout(this, 20) << "INFO: removed entries up to: " << end_marker <<  " from queue: " 
-          << queue_name << dendl;
+          ldpp_dout(this, 20) << "INFO: removed entries up to: " << end_marker <<  " from queue: " << queue_name << dendl;
+        }
+
+        // reserving and committing the migrating entries
+        if (migration_in_process) {
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:471]: you can just check that `entries_to_migrate.size() > 0` and the nyou don't need the `migration_in_process` flag

File: src/rgw/driver/rados/rgw_notify.cc [hunk 10]
```diff
@@ -440,9 +462,67 @@ class Manager : public DoutPrefixProvider {
...
+          std::vector<bufferlist> migration_vector;
+          uint64_t size_to_migrate = 0;
+          RGWPubSub ps(&rados_store, tenant_name);
+
+          rgw_pubsub_topic topic;
+          auto ret_of_get_topic = ps.get_topic(this, queue_name, topic, null_yield);
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:477]: please use `yield` and not `null_yield`

File: src/rgw/driver/rados/rgw_notify.cc [hunk 11]
```diff
@@ -440,9 +462,67 @@ class Manager : public DoutPrefixProvider {
...
+          const auto& conf = this->get_cct()->_conf;
+
+          for (auto entry: entries_to_migrate) {
+            if (ret_of_get_topic < 0) {
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:481]: you can check that after line 477 and print an error message with the matching error code to the debug log

File: src/rgw/driver/rados/rgw_notify.cc [hunk 12]
```diff
@@ -440,9 +462,67 @@ class Manager : public DoutPrefixProvider {
...
+              return;
+            }
+            size_to_migrate += entry.data.length();
+            event_entry_t event_entry;
+            auto iter = entry.data.cbegin();
+            try {
+              decode(event_entry, iter);
+            } catch (buffer::error& err) {
+              ldpp_dout(this, 5) << "WARNING: failed to decode entry. error: " << err.what() << dendl;
+              break;
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:492]: if we failed to decode just one entry, maybe we should use `continue` and not `break`?

File: src/rgw/driver/rados/rgw_notify.cc (hunk 5 above)
Comment (yuvalif) [reply src/rgw/driver/rados/rgw_notify.cc:213]: please extract the tenant name from the queue name

File: src/cls/queue/cls_queue_types.h (hunk 6 above)
Comment (yuvalif) [reply src/cls/queue/cls_queue_types.h:23]: please break to a separate vector of booleans
this is just used in-memoery, no need to encode/decode

Comment (AliMasarweh) [reply src/cls/queue/cls_queue_types.h]: sure

File: src/rgw/driver/rados/rgw_notify.cc (hunk 7 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc]: done

File: src/rgw/driver/rados/rgw_notify.cc (hunk 8 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc]: done!

File: src/rgw/driver/rados/rgw_notify.cc (hunk 9 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc]: done

File: src/rgw/driver/rados/rgw_notify.cc (hunk 10 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc]: done!

File: src/rgw/driver/rados/rgw_notify.cc (hunk 12 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc]: Ack!

File: src/rgw/driver/rados/rgw_notify.cc (hunk 11 above)
Comment (AliMasarweh) [reply src/rgw/driver/rados/rgw_notify.cc:472]: Ack!

File: src/test/rgw/bucket_notification/test_bn.py (hunk 4 above)
Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:3265]: ok. but this was always the case, unrelated to this PR, right?

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
@AliMasarweh did you have a passing teuthology run after the fix?

**Comment by yuvalif** [comment]
please add 2 comments in the bucket notification doc:
* when upgrading to squid, retry limits will start from the time of migration and not the entry creation for  topics created before the migration
* when upgrading to squid, retry limits do not apply to tenanted topics created before the migration

File: src/test/rgw/bucket_notification/test_bn.py (hunk 4 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:3265]: yes this PR didn't change the functionality of retries
